### ISSUE TITLE
Incorrectly inferring generic function parameter type

### ISSUE THREAD

**Issue by Liet3** [microsoft/pyright#10703]
Type: <b>Bug</b>

Hi. I'm using Python3.13. I have code that looks like this and works fine:
```python
from __future__ import annotations
from typing import  Sequence, TypeIs, Callable

def f[B, A](a: Sequence[B | A], g: Callable[[B | A], TypeIs[A]]) -> tuple[B, A]:
    ...

def g(x: num | int) -> TypeIs[int]:
    ...

class num(int):
    pass

l: list[num | int] = [num(1), 3]
h = f(l, g)
```
This works fine, but when I swap the `A` and `B` type variables, I get an error:

```python
from __future__ import annotations
from typing import  Sequence, TypeIs, Callable

def f[A, B](a: Sequence[A | B], g: Callable[[A | B], TypeIs[B]]) -> tuple[A, B]:
    ...

def g(x: num | int) -> TypeIs[int]:
    ...

class num(int):
    pass

l: list[num | int] = [num(1), 3]
h = f(l, g) # Error on g ... Pylance(reportArgumentType)
```
```
Argument of type "(x: num | int) -> TypeIs[int]" cannot be assigned to parameter "g" of type "(A@f | B@f) -> TypeIs[B@f]" in function "f"
  Type "(x: num | int) -> TypeIs[int]" is not assignable to type "(int | num) -> TypeIs[num]"
    Function return type "TypeIs[int]" is incompatible with type "TypeIs[num]"
      "TypeIs[int]" is not assignable to "TypeIs[num]"
        Type parameter "T@TypeIs" is invariant, but "int" is not the same as "num"
      "bool" is not assignable to "TypeIs[num]"Pylance(reportArgumentType)
```
even though these are functionally the same.

Extension version: 2025.6.2
VS Code version: Code 1.102.0 (cb0c47c0cfaad0757385834bd89d410c78a856c0, 2025-07-09T22:10:34.600Z)
OS version: Windows_NT x64 10.0.19045
Modes:

<details>
<summary>System Added Logs</summary>

```

from __future__ import annotations
from typing import  Sequence, TypeIs, Callable


def f[A, B](a: Sequence[A | B], g: Callable[[A | B], TypeIs[B]]) -> tuple[A, B]:
    ...

def g(x: num | int) -> TypeIs[int]:
    ...

class num(int):
    pass

l: list[num | int] = [num(1), 3]
h = f(l, g) # Error ... Pylance(reportArgumentType)




```
</details><details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|AMD Ryzen 7 4800H with Radeon Graphics          (16 x 2894)|
|GPU Status|2d_canvas: enabled<br>canvas_oop_rasterization: enabled_on<br>direct_rendering_display_compositor: disabled_off_ok<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>opengl: enabled_on<br>rasterization: enabled<br>raw_draw: disabled_off_ok<br>skia_graphite: disabled_off<br>video_decode: enabled<br>video_encode: enabled<br>vulkan: disabled_off<br>webgl: enabled<br>webgl2: enabled<br>webgpu: enabled<br>webnn: disabled_off|
|Load (avg)|undefined|
|Memory (System)|31.37GB (17.31GB free)|
|Process Argv|--crash-reporter-id 0a2c6f32-27fc-4931-a362-baccdd1378e2|
|Screen Reader|no|
|VM|0%|
</details><details>
<summary>A/B Experiments</summary>

```
vsliv368:30146709
vspor879:30202332
vspor708:30202333
vspor363:30204092
binariesv615:30325510
c4g48928:30535728
azure-dev_surveyone:30548225
962ge761:30959799
h48ei257:31000450
cppperfnew:31000557
dwnewjupytercf:31046870
nativeloc1:31344060
5fd0e150:31155592
dwcopilot:31170013
6074i472:31201624
dwoutputs:31242946
9064b325:31222308
copilot_t_ci:31333650
e5gg6876:31282496
pythoneinst12:31285622
c7cif404:31314491
pythonpulldiag:31343502
996jf627:31283433
pythonrdcb7:31342333
usemplatestapi:31297334
0aa6g176:31307128
747dc170:31275177
aj953862:31281341
generatesymbolt:31295002
convertfstringf:31295003
pylancequickfixt:31329274
9d2cg352:31346308
usemarketplace:31343026
nesew2to5:31336538
agentclaude:31335814
b6b4d950:31327385
nes-diff-11:31337487
nes-set-on:31340697
6abeh943:31336334
yijiwantestdri0626-c:31336931
64bie973:31346034
f76d9909:31342392

```

</details>

**Comment by erictraut** [comment]
Here's what's going on. When a type checker encounters a call to a generic function, it needs to "solve" the type variables in the call's signature. It does this by first establishing "constraints" for the type variable. These constraints come from the types of the supplied arguments. Once these constraints are collected, it then "solves" the constraints for each type variable. 

Let's walk through a simple example first.

```python
def func[T](x: T, y: T) -> T: ...

func(1, "")
```

In this simple example, there are two constraints imposed on `T`. The first constraint is `T >: int` (which you can read as "`T` must at least as wide as type `int`). This constraint comes from the first argument `1`. A second constraint (`T >: str`) comes from the second argument `""`.  These constraints use the `>:` operator instead of `=:` (which you can read as "`T` must be equivalent to `int`) because this is a covariant rather than invariant context. There are many solutions for these two constraints. Mypy chooses the solution `T := object`, whereas pyright chooses the solution `T := int | str`. Both are valid.

Now, let's look at your example, which makes use of a union of two type variables. This situation creates ambiguities for a constraint solver and can lead to behaviors like the one you're observing. The type of argument `l` is `list[num | int]`, and this value is being assigned to parameter `a` whose type is `Sequence[B | A]`. Which constraints should this impose on `A` and `B`? There are different ways to answer this question. For example, we could impose the constraint `B >: num | int` with no constraint on `A`. Or we could impose the constraints `B >: num` and `A >: int`. Or we could impose the constraints `B >: int` and `A >: num`. (The situation would be even more complicated if the source union included three or more subtypes (e.g. `num | int | str | None`).) Pyright uses a variety of heuristics to pick the constraints that are most likely to align with the intent of the developer. It also attempts to do this in a way that is deterministic — that doesn't depend on parameter or argument order or the order in which subtypes are specified within a union. To accomplish this, it internally normalizes all unions by sorting its subtypes. For example, the unions `B | A` and `A | B` are normalized internally to `A | B`, and `int | num` and `num | int` are normalized to `int | num`. When pyright sees that one union is assigned to another union and the destination contains multiple type variables, it attempts to associate constraints with each of the type variables. This heuristic works well for most use cases.

Now that you understand the algorithm, let's look at your second example — the one that produces an error. Here, the normalized type of argument `l` is `list[int | num]`, and the normalized type of argument `g` is `(x: int | num) -> TypeIs[int]`. The following constraints are established for the call `f(l, g)`:

```
# Constraints from parameter a
A >: int  # Covariant context
B >: num  # Covariant context

# Constraints from parameter g
A <: int  # Contravariant context
B <: num  # Contravariant context
B =: int  # Invariant context
```

Note that the `B =: int` constraint comes from the `TypeIs` type. Its type parameter is invariant, which explains the `:=` here. For details, refer to [this chapter](https://typing.python.org/en/latest/spec/narrowing.html#typeis) of the typing spec. Note also that we picked up some contravariant constraints because of the `Callable` type. All input parameters to a Callable are implicitly contravariant.

If you look closely at this set of constraints, you can see that there is valid solution. In particular, There is no solution for `B` that satisfies all three constraints that involve `B`.

Why does this work when you rename `A` to `B` and vice versa? Let's walk through that example as well.

Here are the constraints that are established in this case:

```
# Constraints from parameter a
A >: int  # Covariant context
B >: num  # Covariant context

# Constraints from parameter g
A <: int  # Contravariant context
B <: num  # Contravariant context
A =: int  # Invariant context
```

These constraints do have a valid solution: `A := int` and `B := num`.


Theoretically, a type checker could attempt to generate all possible combinations of constraint sets and try to solve all of them in turn, but in practice this wouldn't be practical. Constraint solving is already one of the most expensive and complex aspects of type checking, and exploring all combinations of potential constraint sets would be exponentially more costly. While the current behavior in this particular case is less than ideal, I don't see a practical way to improve it.

I'll keep the issue open and think about it a bit more, but I will likely close this at some point in the future.

Let me know if you have any questions about the above analysis.

**Comment by L-iet** [comment]
Commenting from my other profile - I see. I had a feeling something like sorting was happening. The error goes away if I add an explicit type annotation:
```python
h: tuple[num, int] = f(l, g)
```
I wonder if there would be a separate diagnostic for when it has not fully explored all possible solutions when constraint solving? Perhaps a message to suggest adding more type annotations. Also maybe letting the user set the depth for constraint solving?
